components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt

```diff
@@ -0,0 +1,557 @@
+            matrix.setRectToRect(viewRect, bufferRect, Matrix.ScaleToFit.FILL)
+            val scale = Math.max(viewHeight.toFloat() / size.height, viewWidth.toFloat() / size.width)
+            matrix.postScale(scale, scale, centerX, centerY)
+            matrix.postRotate((90 * (rotation - 2)).toFloat(), centerX, centerY)
+        } else if (Surface.ROTATION_180 == rotation) {
+            matrix.postRotate(180f, centerX, centerY)
+        }
+        textureView.setTransform(matrix)
+    }
+
+    /**
+     * Creates a new [CameraCaptureSession] for camera preview.
+     */
+    private fun createCameraPreviewSession() {
+        val texture = textureView.surfaceTexture
+
+        val size = previewSize as Size
+        // We configure the size of default buffer to be the size of camera preview we want.
+        texture.setDefaultBufferSize(size.width, size.height)
+
+        val surface = Surface(texture)
+        val mImageSurface = imageReader?.surface
+        cameraDevice?.let {
+            previewRequestBuilder = it.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW).apply {
+                addTarget(mImageSurface as Surface)
+                addTarget(surface)
+            }
+
+            it.createCaptureSession(Arrays.asList(mImageSurface, surface),
+                object : CameraCaptureSession.StateCallback() {
+
+                    override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
+                        if (null == cameraDevice) return
+
+                        previewRequestBuilder?.set(CaptureRequest.CONTROL_AF_MODE,
+                                CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
+
+                        previewRequest = previewRequestBuilder?.build()
+
+                        captureSession = cameraCaptureSession
+                        cameraCaptureSession.setRepeatingRequest(previewRequest as CaptureRequest,
+                            object : CameraCaptureSession.CaptureCallback() {},
+                            backgroundHandler
+                        )
+                    }
+
+                    override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
+                        Log.log(Log.Priority.ERROR, "mozac-qr", message = "Failed to configure CameraCaptureSession")
+                    }
+                }, null
+            )
+        }
+    }
+
+    /**
+     * Compares two `Size`s based on their areas.
+     */
+    internal class CompareSizesByArea : Comparator<Size> {
+        override fun compare(lhs: Size, rhs: Size): Int {
+            return java.lang.Long.signum(lhs.width.toLong() * lhs.height - rhs.width.toLong() * rhs.height)
+        }
+    }
+
+    companion object {
+        fun newInstance(listener: OnScanCompleteListener): QrFragment {
+            val fragment = QrFragment()
+            val args = Bundle().apply {
+                putSerializable(BUNDLE_SCAN_COMPLETE_LISTENER, listener)
+            }
```

Comment:
This part seems problematic to me. The listener is a function which keeps a reference to whatever outside of the feature. Technically this can't really be serialized?

In other similar situations we:
* Create the fragment and set a reference to the feature on the fragment
* Since Android re-creates fragments when the app was destroyed in the background we handle this in the feature: If on initialization there's already a fragment in the fragment manager then we re-set the reference to the new feature.

See `DownloadsFeature` as an example.

Reply:
@pocmo ah, this handles a different case. I should've commented. When the orientation of device changes while scanning, the fragment gets re-created and the listener would be lost otherwise. I tried a few options but only made that case work with this "workaround".

Reply:
But that's the same case we handle in `DownloadsFeature`? I can't see how this can work without leaking everything that is referenced from the callback. And I'm confused how this can work when the process gets re-created?

Reply:
I will try this again (re-using the fragment instance instead.)

Reply:
OK, this is fixed now!